README.md

```diff
@@ -1,6 +1,93 @@
+
+## Implementation Challenges
+
+There are four pillars to Look.
+
+### Linearizing Vector Clocks
+
+In terms of functional programming, the inputs to Look are monadic transormations. Transactional Outboxes, Kafka consumers, logical replication logs, CDC pipelines and Git commits can all be viewed as such.
+
+The inputs do not have to be synchronized with one another; even Kafka partitions within the same topic are not synchronized. Thus, the spectrum of inputs to a logical instance of Look can be described by a vector clock. Inside Look, data updates are applied in a strictly sequential, totally ordered, linearized way.
+
+Therefore, for each particular incoming event from any of the Logical Transactional Outbox, a unique sequence ID number (the epoch number) is stamped.
+
+Inversely, each epoch number within a logical Look instance corresponds to one atomic update in the lifetime of this Look; most often an update to the data it stores, although system-level updates, such as changes to data schema, its processing logic, or deployment topology, also get their own unique epoch numbers, interleaved with data updates in a totally ordered fashion.
+
+### In-Memory Event Store
+
+Under the hood, Look maintains a complete event log of all the mutations that took place, both to the data and to the logic that handles this data. Thus, it is accurate to say that Look is based on an immutable, append-only event log.
+
+The in-memory storage of Look follows the approach of [persistent data structures](https://en.wikipedia.org/wiki/Persistent_data_structure), the nodes of which are compactified and serialized to disk. The storage keeps the "dirtied" pages as well, on disk for the history, as well as in memory, for a configurable amount of wall time and epoch number; some 15 seconds and a million mutations by default.
+
+To a certain degree, this implies that read queries that refer to a specific epoch number are fully idempotent, as long as they are performed against a replica that is configured to store enough of the past state. It is also possible to spin a dedicated replica that is configured to keep more of historic data in hot storage. The storage layer of Look also performs regular snapshots, so, for reconciliation reasons, it is possible to "roll it back", temporarily and on a dedicated node or set of nodes, to the state of data & code that was actual far away in the past.
+
+Last but not least: adding Look replicas is a straightforward task. For read-only replicas, it is as trivial as spinning up a node and pointing it at the currently up and running Look cluster. For read-write replicas, a.k.a. potential leaders when it comes to the high availability and durability of the distributed system, the node is first added as read-only, and, as it has caught up, the record that registers it as such can be added into the cluster configuration namespace of the immutable append-only log of this Look deployment.
```

Comment:
So what will be our replication model for the storage? I see read-write replicas. Does it mean that we will not follow with a usual leader-based approach?